IllegalStateException in HttpOutput with Jersey

**Jetty version**
9.4.25.v20191220

**Java version**
OpenJDK 1.8.0_232 (Zulu)

**OS type/version**
macOS Mojave 10.14.6

**Description**
Jetty 9.4.25 seems to have a regression (or stricter behavior) in the `HttpOutput` state machine (likely introduced by #4409) which causes the following exception on `write()`:
```
java.lang.IllegalStateException: s=OPEN,api=BLOCKED,sc=false,e=null
```
Unfortunately, I was not able to create a simple reproduction, but it reproduces reliably when running a Presto unit test in the `presto-jmx` module from this branch: https://github.com/electrum/presto/tree/jetty9425:
```
mvn clean install -DskipTests -pl presto-jmx -am -T1C
mvn test -pl presto-jmx -Dtest=io.prestosql.plugin.jmx.TestJmxQueries#testQuery
```
After running the above, look for the following error in the test output after 5-10 seconds. You can kill the test at this point, since it takes a while to eventually time out.
```
2020-01-07T14:01:47.723-0600 SEVERE An I/O error has occurred while writing a response message entity to the container output stream.
org.glassfish.jersey.server.internal.process.MappableException: com.fasterxml.jackson.databind.JsonMappingException: s=OPEN,api=BLOCKED,sc=false,e=null (through reference chain: io.prestosql.client.QueryResults["data"]->java.util.Collections$UnmodifiableRandomAccessList[12])
	at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:91)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:163)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1135)
	at org.glassfish.jersey.server.ServerRuntime$Responder.writeResponse(ServerRuntime.java:662)
	at org.glassfish.jersey.server.ServerRuntime$Responder.processResponse(ServerRuntime.java:395)
	at org.glassfish.jersey.server.ServerRuntime$Responder.process(ServerRuntime.java:385)
	at org.glassfish.jersey.server.ServerRuntime$AsyncResponder$3.run(ServerRuntime.java:884)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:272)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:268)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:316)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:298)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:268)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:289)
	at org.glassfish.jersey.server.ServerRuntime$AsyncResponder.resume(ServerRuntime.java:916)
	at org.glassfish.jersey.server.ServerRuntime$AsyncResponder.resume(ServerRuntime.java:872)
	at io.airlift.http.server.AsyncResponseHandler$1.onSuccess(AsyncResponseHandler.java:102)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1058)
	at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:78)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: s=OPEN,api=BLOCKED,sc=false,e=null (through reference chain: io.prestosql.client.QueryResults["data"]->java.util.Collections$UnmodifiableRandomAccessList[12])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:394)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:365)
	at com.fasterxml.jackson.databind.ser.std.StdSerializer.wrapAndThrow(StdSerializer.java:338)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:123)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
	at com.fasterxml.jackson.databind.ser.std.IterableSerializer.serializeContents(IterableSerializer.java:107)
	at com.fasterxml.jackson.databind.ser.std.IterableSerializer.serialize(IterableSerializer.java:74)
	at com.fasterxml.jackson.databind.ser.std.IterableSerializer.serialize(IterableSerializer.java:12)
	at com.fasterxml.jackson.databind.ser.std.StdDelegatingSerializer.serialize(StdDelegatingSerializer.java:168)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:727)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:721)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:166)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1433)
	at com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:921)
	at io.airlift.jaxrs.JsonMapper.writeTo(JsonMapper.java:233)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.invokeWriteTo(WriterInterceptorExecutor.java:266)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.aroundWriteTo(WriterInterceptorExecutor.java:251)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:163)
	at org.glassfish.jersey.server.internal.JsonWithPaddingInterceptor.aroundWriteTo(JsonWithPaddingInterceptor.java:109)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:163)
	at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:85)
	... 20 more
Caused by: java.lang.IllegalStateException: s=OPEN,api=BLOCKED,sc=false,e=null
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:767)
	at io.airlift.http.server.TimingFilter$TimedServletOutputStream.write(TimingFilter.java:163)
	at org.glassfish.jersey.servlet.internal.ResponseWriter$NonCloseableOutputStreamWrapper.write(ResponseWriter.java:325)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.write(CommittingOutputStream.java:224)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$UnCloseableOutputStream.write(WriterInterceptorExecutor.java:300)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._flushBuffer(UTF8JsonGenerator.java:2136)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._writeCustomStringSegment2(UTF8JsonGenerator.java:1657)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._writeStringSegment(UTF8JsonGenerator.java:1396)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._writeStringSegments(UTF8JsonGenerator.java:1281)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator.writeString(UTF8JsonGenerator.java:502)
	at com.fasterxml.jackson.databind.ser.std.StringSerializer.serialize(StringSerializer.java:41)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
	... 40 more
```

same here but with other frameworks (spring boot 2.2)

```
2020-01-08 11:27:17.629  WARN 14955 --- [qtp509559152-14,] org.eclipse.jetty.server.HttpChannel     : /vendor-es2015.ab355fce23594e19a032.js
org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.IllegalStateException: s=OPEN,api=BLOCKED,sc=false,e=null
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:503)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
        at XXX.web.filter.CachingHttpHeadersFilter.doFilter(CachingHttpHeadersFilter.java:52)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
        at XXX.web.filter.RequestAndResponseLoggingFilter.doFilterWrapped(RequestAndResponseLoggingFilter.java:65)
        at XXX.web.filter.RequestAndResponseLoggingFilter.doFilterInternal(RequestAndResponseLoggingFilter.java:49)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
        at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:92)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126)
        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Caused by: java.lang.IllegalStateException: s=OPEN,api=BLOCKED,sc=false,e=null
        at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:767)
        at org.springframework.security.web.util.OnCommittedResponseWrapper$SaveContextServletOutputStream.write(OnCommittedResponseWrapper.java:645)
        at org.springframework.util.StreamUtils.copy(StreamUtils.java:143)
        at org.springframework.http.converter.ResourceHttpMessageConverter.writeContent(ResourceHttpMessageConverter.java:132)
        at org.springframework.http.converter.ResourceHttpMessageConverter.writeInternal(ResourceHttpMessageConverter.java:124)
        at org.springframework.http.converter.ResourceHttpMessageConverter.writeInternal(ResourceHttpMessageConverter.java:45)
        at org.springframework.http.converter.AbstractHttpMessageConverter.write(AbstractHttpMessageConverter.java:227)
        at org.springframework.web.servlet.resource.ResourceHttpRequestHandler.handleRequest(ResourceHttpRequestHandler.java:489)
        at org.springframework.web.servlet.mvc.HttpRequestHandlerAdapter.handle(HttpRequestHandlerAdapter.java:53)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:503)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
        at XXX.web.filter.CachingHttpHeadersFilter.doFilter(CachingHttpHeadersFilter.java:52)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
```

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
@@ -1696,16 +1696,6 @@ public class HttpOutput extends ServletOutputStream implements Runnable
                 return Action.SCHEDULED;
             }
 
-            // all content written, but if we have not yet signaled completion,
-            // then we need to do so
-            if (_last && !_completed)
-            {
-                // TODO How can this ever happen?
-                _completed = true;
-                channelWrite(BufferUtil.EMPTY_BUFFER, true, this);
-                return Action.SCHEDULED;
-            }
-
             if (LOG.isDebugEnabled() && _completed)
                 LOG.debug("EOF of {}", this);
 

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
@@ -1628,7 +1628,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         private final ByteBuffer _buffer;
         private final ByteBuffer _slice;
         private final int _len;
-        volatile boolean _completed;
+        private boolean _completed;
 
         AsyncWrite(byte[] b, int off, int len, boolean last)
         {
@@ -1696,10 +1696,11 @@ public class HttpOutput extends ServletOutputStream implements Runnable
                 return Action.SCHEDULED;
             }
 
-            // all content written, but if we have not yet signal completion, we
-            // need to do so
+            // all content written, but if we have not yet signaled completion,
+            // then we need to do so
             if (_last && !_completed)
             {
+                // TODO How can this ever happen?
                 _completed = true;
                 channelWrite(BufferUtil.EMPTY_BUFFER, true, this);
                 return Action.SCHEDULED;

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
@@ -378,6 +378,19 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         return wake;
     }
 
+    private int getAggregateSpace()
+    {
+        // If no aggregate, we can allocate one of bufferSize
+        if (_aggregate == null)
+            return getBufferSize();
+
+        // if the position is not zero, compact to avoid empty at capacity
+        if (_aggregate.position() != 0)
+            BufferUtil.compact(_aggregate);
+
+        return BufferUtil.space(_aggregate);
+    }
+
     public void softClose()
     {
         synchronized (_channelState)
@@ -723,6 +736,9 @@ public class HttpOutput extends ServletOutputStream implements Runnable
     @Override
     public void write(byte[] b, int off, int len) throws IOException
     {
+        if (LOG.isDebugEnabled())
+            LOG.debug("write(array {})", BufferUtil.toDetailString(ByteBuffer.wrap(b, off, len)));
+
         boolean last;
         boolean aggregate;
         boolean flush;
@@ -733,7 +749,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         {
             checkWritable();
             long written = _written + len;
-            int space = _aggregate == null ? getBufferSize() : BufferUtil.space(_aggregate);
+            int space = getAggregateSpace();
             last = _channel.getResponse().isAllContentWritten(written);
             // Write will be aggregated if:
             //  + it is smaller than the commitSize
@@ -777,13 +793,22 @@ public class HttpOutput extends ServletOutputStream implements Runnable
 
                 // return if we are not complete, not full and filled all the content
                 if (!flush)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("write(array) {} aggregated !flush {}",
+                            stateString(), BufferUtil.toDetailString(_aggregate));
                     return;
+                }
 
                 // adjust offset/length
                 off += filled;
                 len -= filled;
             }
         }
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("write(array) {} last={} agg={} flush=true async={}, len={} {}",
+                stateString(), last, aggregate, async, len, BufferUtil.toDetailString(_aggregate));
 
         if (async)
         {
@@ -801,9 +826,10 @@ public class HttpOutput extends ServletOutputStream implements Runnable
                 channelWrite(_aggregate, last && len == 0);
 
                 // should we fill aggregate again from the buffer?
-                if (len > 0 && !last && len <= _commitSize && len <= BufferUtil.space(_aggregate))
+                if (len > 0 && !last && len <= _commitSize && len <= getAggregateSpace())
                 {
                     BufferUtil.append(_aggregate, b, off, len);
+                    onWriteComplete(false, null);
                     return;
                 }
             }
@@ -929,7 +955,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         {
             checkWritable();
             long written = _written + 1;
-            int space = _aggregate == null ? getBufferSize() : BufferUtil.space(_aggregate);
+            int space = getAggregateSpace();
             last = _channel.getResponse().isAllContentWritten(written);
             flush = last || space == 1;
 
@@ -1639,7 +1665,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
             }
 
             // Can we just aggregate the remainder?
-            if (!_last && _len < BufferUtil.space(_aggregate) && _len < _commitSize)
+            if (!_last && _aggregate != null && _len < getAggregateSpace() && _len < _commitSize)
             {
                 int position = BufferUtil.flipToFill(_aggregate);
                 BufferUtil.put(_buffer, _aggregate);

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java
@@ -378,15 +378,14 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         return wake;
     }
 
-    private int getAggregateSpace()
+    private int maximizeAggregateSpace()
     {
         // If no aggregate, we can allocate one of bufferSize
         if (_aggregate == null)
             return getBufferSize();
 
-        // if the position is not zero, compact to avoid empty at capacity
-        if (_aggregate.position() != 0)
-            BufferUtil.compact(_aggregate);
+        // compact to maximize space
+        BufferUtil.compact(_aggregate);
 
         return BufferUtil.space(_aggregate);
     }
@@ -749,7 +748,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         {
             checkWritable();
             long written = _written + len;
-            int space = getAggregateSpace();
+            int space = maximizeAggregateSpace();
             last = _channel.getResponse().isAllContentWritten(written);
             // Write will be aggregated if:
             //  + it is smaller than the commitSize
@@ -826,7 +825,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
                 channelWrite(_aggregate, last && len == 0);
 
                 // should we fill aggregate again from the buffer?
-                if (len > 0 && !last && len <= _commitSize && len <= getAggregateSpace())
+                if (len > 0 && !last && len <= _commitSize && len <= maximizeAggregateSpace())
                 {
                     BufferUtil.append(_aggregate, b, off, len);
                     onWriteComplete(false, null);
@@ -955,7 +954,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
         {
             checkWritable();
             long written = _written + 1;
-            int space = getAggregateSpace();
+            int space = maximizeAggregateSpace();
             last = _channel.getResponse().isAllContentWritten(written);
             flush = last || space == 1;
 
@@ -1665,7 +1664,7 @@ public class HttpOutput extends ServletOutputStream implements Runnable
             }
 
             // Can we just aggregate the remainder?
-            if (!_last && _aggregate != null && _len < getAggregateSpace() && _len < _commitSize)
+            if (!_last && _aggregate != null && _len < maximizeAggregateSpace() && _len < _commitSize)
             {
                 int position = BufferUtil.flipToFill(_aggregate);
                 BufferUtil.put(_buffer, _aggregate);
@@ -1693,6 +1692,15 @@ public class HttpOutput extends ServletOutputStream implements Runnable
                 _slice.position(p);
                 _completed = !_buffer.hasRemaining();
                 channelWrite(_slice, _last && _completed, this);
+                return Action.SCHEDULED;
+            }
+
+            // all content written, but if we have not yet signal completion, we
+            // need to do so
+            if (_last && !_completed)
+            {
+                _completed = true;
+                channelWrite(BufferUtil.EMPTY_BUFFER, true, this);
                 return Action.SCHEDULED;
             }
 
@@ -1863,32 +1871,16 @@ public class HttpOutput extends ServletOutputStream implements Runnable
 
     private class WriteCompleteCB implements Callback
     {
-        final Callback _callback;
-
-        WriteCompleteCB()
-        {
-            this(null);
-        }
-
-        WriteCompleteCB(Callback callback)
-        {
-            _callback = callback;
-        }
-
         @Override
         public void succeeded()
         {
             onWriteComplete(true, null);
-            if (_callback != null)
-                _callback.succeeded();
         }
 
         @Override
         public void failed(Throwable x)
         {
             onWriteComplete(true, x);
-            if (_callback != null)
-                _callback.succeeded();
         }
     }
 }

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHttpOutputInterceptor.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHttpOutputInterceptor.java
@@ -381,7 +381,7 @@ public class GzipHttpOutputInterceptor implements HttpOutput.Interceptor
                         int len = slice.remaining();
                         _crc.update(array, off, len);
                         _deflater.setInput(array, off, len);  // TODO use ByteBuffer API in Jetty-10
-                        BufferUtil.clear(slice);
+                        slice.position(slice.position() + len);
                         if (_last && BufferUtil.isEmpty(_content))
                             _deflater.finish();
                     }

--- a/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
+++ b/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
@@ -826,29 +826,6 @@ public class HttpOutputTest
             baseRequest.setHandled(true);
             HttpOutput out = (HttpOutput) response.getOutputStream();
 
-            // Add clearing interceptor to simulate behaviour of GzipHandler
-            HttpOutput.Interceptor interceptor = out.getInterceptor();
-            out.setInterceptor(new Interceptor()
-            {
-                @Override
-                public void write(ByteBuffer content, boolean last, Callback callback)
-                {
-                    interceptor.write(content, last, Callback.from(()->BufferUtil.clear(content), callback));
-                }
-
-                @Override
-                public Interceptor getNextInterceptor()
-                {
-                    return interceptor;
-                }
-
-                @Override
-                public boolean isOptimizedForDirectBuffers()
-                {
-                    return interceptor.isOptimizedForDirectBuffers();
-                }
-            });
-
             int bufferSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputBufferSize();
             int commitSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputAggregationSize();
             char fill = 'A';

--- a/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
+++ b/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
@@ -18,6 +18,7 @@
 
 package org.eclipse.jetty.server;
 
+import java.io.ByteArrayOutputStream;
 import java.io.FilterInputStream;
 import java.io.IOException;
 import java.io.InputStream;
@@ -54,6 +55,8 @@ import static org.junit.jupiter.api.Assertions.assertTrue;
  */
 public class HttpOutputTest
 {
+    public static final int OUTPUT_AGGREGATION_SIZE = 1024;
+    public static final int OUTPUT_BUFFER_SIZE = 4096;
     private Server _server;
     private LocalConnector _connector;
     private ContentHandler _handler;
@@ -67,7 +70,8 @@ public class HttpOutputTest
         HttpConnectionFactory http = new HttpConnectionFactory();
         http.getHttpConfiguration().setRequestHeaderSize(1024);
         http.getHttpConfiguration().setResponseHeaderSize(1024);
-        http.getHttpConfiguration().setOutputBufferSize(4096);
+        http.getHttpConfiguration().setOutputBufferSize(OUTPUT_BUFFER_SIZE);
+        http.getHttpConfiguration().setOutputAggregationSize(OUTPUT_AGGREGATION_SIZE);
 
         _connector = new LocalConnector(_server, http, null);
         _server.addConnector(_connector);
@@ -676,6 +680,205 @@ public class HttpOutputTest
         assertThat(response, containsString("400\tTHIS IS A BIGGER FILE"));
     }
 
+    @Test
+    public void testAggregation() throws Exception
+    {
+        AggregateHandler handler = new AggregateHandler();
+        _swap.setHandler(handler);
+        handler.start();
+        String response = _connector.getResponse("GET / HTTP/1.0\nHost: localhost:80\n\n");
+        assertThat(response, containsString("HTTP/1.1 200 OK"));
+        assertThat(response, containsString(handler.expected.toString()));
+    }
+
+    static class AggregateHandler extends AbstractHandler
+    {
+        ByteArrayOutputStream expected = new ByteArrayOutputStream();
+        @Override
+        public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+        {
+            baseRequest.setHandled(true);
+            HttpOutput out = (HttpOutput) response.getOutputStream();
+
+            // Add interceptor to check aggregation is done
+            HttpOutput.Interceptor interceptor = out.getInterceptor();
+            out.setInterceptor(new AggregationChecker(interceptor));
+
+            int bufferSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputBufferSize();
+            int len = bufferSize * 3 / 2;
+
+            byte[] data = new byte[AggregationChecker.MAX_SIZE];
+            int fill = 0;
+            while (expected.size() < len)
+            {
+                Arrays.fill(data, (byte)('A' + (fill++%26)));
+                expected.write(data);
+                out.write(data);
+            }
+        }
+    }
+
+    @Test
+    public void testAsyncAggregation() throws Exception
+    {
+        AsyncAggregateHandler handler = new AsyncAggregateHandler();
+        _swap.setHandler(handler);
+        handler.start();
+        String response = _connector.getResponse("GET / HTTP/1.0\nHost: localhost:80\n\n");
+        assertThat(response, containsString("HTTP/1.1 200 OK"));
+        assertThat(response, containsString(handler.expected.toString()));
+    }
+
+    static class AsyncAggregateHandler extends AbstractHandler
+    {
+        ByteArrayOutputStream expected = new ByteArrayOutputStream();
+        @Override
+        public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+        {
+            baseRequest.setHandled(true);
+            HttpOutput out = (HttpOutput) response.getOutputStream();
+
+            // Add interceptor to check aggregation is done
+            HttpOutput.Interceptor interceptor = out.getInterceptor();
+            out.setInterceptor(new AggregationChecker(interceptor));
+
+            int bufferSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputBufferSize();
+            int len = bufferSize * 3 / 2;
+
+            AsyncContext async = request.startAsync();
+            out.setWriteListener(new WriteListener()
+            {
+                int fill = 0;
+                @Override
+                public void onWritePossible() throws IOException
+                {
+                    byte[] data = new byte[AggregationChecker.MAX_SIZE];
+                    while(out.isReady())
+                    {
+                        if (expected.size() >= len)
+                        {
+                            async.complete();
+                            return;
+                        }
+
+                        Arrays.fill(data, (byte)('A' + (fill++%26)));
+                        expected.write(data);
+                        out.write(data);
+                    }
+                }
+
+                @Override
+                public void onError(Throwable t)
+                {
+                }
+            });
+        }
+    }
+
+    private static class AggregationChecker implements Interceptor
+    {
+        static final int MAX_SIZE = OUTPUT_AGGREGATION_SIZE / 2 - 1;
+        private final Interceptor interceptor;
+
+        public AggregationChecker(Interceptor interceptor)
+        {
+            this.interceptor = interceptor;
+        }
+
+        @Override
+        public void write(ByteBuffer content, boolean last, Callback callback)
+        {
+            if (content.remaining() <= MAX_SIZE)
+                throw new IllegalStateException("Not Aggregated!");
+            interceptor.write(content, last, callback);
+        }
+
+        @Override
+        public Interceptor getNextInterceptor()
+        {
+            return interceptor;
+        }
+
+        @Override
+        public boolean isOptimizedForDirectBuffers()
+        {
+            return interceptor.isOptimizedForDirectBuffers();
+        }
+    }
+
+    @Test
+    public void testAggregateResidue() throws Exception
+    {
+        AggregateResidueHandler handler = new AggregateResidueHandler();
+        _swap.setHandler(handler);
+        handler.start();
+        String response = _connector.getResponse("GET / HTTP/1.0\nHost: localhost:80\n\n");
+        assertThat(response, containsString("HTTP/1.1 200 OK"));
+        assertThat(response, containsString(handler.expected.toString()));
+    }
+
+    static class AggregateResidueHandler extends AbstractHandler
+    {
+        ByteArrayOutputStream expected = new ByteArrayOutputStream();
+        @Override
+        public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+        {
+            baseRequest.setHandled(true);
+            HttpOutput out = (HttpOutput) response.getOutputStream();
+
+            // Add clearing interceptor to simulate behaviour of GzipHandler
+            HttpOutput.Interceptor interceptor = out.getInterceptor();
+            out.setInterceptor(new Interceptor()
+            {
+                @Override
+                public void write(ByteBuffer content, boolean last, Callback callback)
+                {
+                    interceptor.write(content, last, Callback.from(()->BufferUtil.clear(content), callback));
+                }
+
+                @Override
+                public Interceptor getNextInterceptor()
+                {
+                    return interceptor;
+                }
+
+                @Override
+                public boolean isOptimizedForDirectBuffers()
+                {
+                    return interceptor.isOptimizedForDirectBuffers();
+                }
+            });
+
+            int bufferSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputBufferSize();
+            int commitSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputAggregationSize();
+            char fill = 'A';
+
+            // write data that will be aggregated
+            byte[] data = new byte[commitSize - 1];
+            Arrays.fill(data, (byte)(fill++));
+            expected.write(data);
+            out.write(data);
+            int aggregated = data.length;
+
+            // write data that will almost fill the aggregate buffer
+            while (aggregated < (bufferSize - 1))
+            {
+                data = new byte[Math.min(commitSize - 1, bufferSize - aggregated - 1)];
+                Arrays.fill(data, (byte)(fill++));
+                expected.write(data);
+                out.write(data);
+                aggregated += data.length;
+            }
+
+            // write data that will not be aggregated
+            data = new byte[bufferSize + 1];
+            Arrays.fill(data, (byte)(fill++));
+            expected.write(data);
+            out.write(data);
+        }
+    }
+
+
     private static String toUTF8String(Resource resource)
         throws IOException
     {
@@ -687,8 +890,6 @@ public class HttpOutputTest
         default void init(Request baseRequest)
         {
         }
-
-        ;
 
         void setNext(Interceptor interceptor);
     }

--- a/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
+++ b/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
@@ -22,8 +22,10 @@ import java.io.ByteArrayOutputStream;
 import java.io.FilterInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.PrintWriter;
 import java.nio.ByteBuffer;
 import java.nio.channels.ReadableByteChannel;
+import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
 import java.util.concurrent.atomic.AtomicInteger;
 import javax.servlet.AsyncContext;
@@ -32,6 +34,7 @@ import javax.servlet.WriteListener;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+import org.eclipse.jetty.io.ByteBufferPool;
 import org.eclipse.jetty.server.HttpOutput.Interceptor;
 import org.eclipse.jetty.server.LocalConnector.LocalEndPoint;
 import org.eclipse.jetty.server.handler.AbstractHandler;
@@ -66,6 +69,20 @@ public class HttpOutputTest
     public void init() throws Exception
     {
         _server = new Server();
+
+        _server.addBean(new ByteBufferPool()
+        {
+            @Override
+            public ByteBuffer acquire(int size, boolean direct)
+            {
+                return direct ? BufferUtil.allocateDirect(size) : BufferUtil.allocate(size);
+            }
+
+            @Override
+            public void release(ByteBuffer buffer)
+            {
+            }
+        });
 
         HttpConnectionFactory http = new HttpConnectionFactory();
         http.getHttpConfiguration().setRequestHeaderSize(1024);
@@ -694,11 +711,12 @@ public class HttpOutputTest
     static class AggregateHandler extends AbstractHandler
     {
         ByteArrayOutputStream expected = new ByteArrayOutputStream();
+
         @Override
         public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
         {
             baseRequest.setHandled(true);
-            HttpOutput out = (HttpOutput) response.getOutputStream();
+            HttpOutput out = (HttpOutput)response.getOutputStream();
 
             // Add interceptor to check aggregation is done
             HttpOutput.Interceptor interceptor = out.getInterceptor();
@@ -711,7 +729,7 @@ public class HttpOutputTest
             int fill = 0;
             while (expected.size() < len)
             {
-                Arrays.fill(data, (byte)('A' + (fill++%26)));
+                Arrays.fill(data, (byte)('A' + (fill++ % 26)));
                 expected.write(data);
                 out.write(data);
             }
@@ -732,11 +750,12 @@ public class HttpOutputTest
     static class AsyncAggregateHandler extends AbstractHandler
     {
         ByteArrayOutputStream expected = new ByteArrayOutputStream();
+
         @Override
         public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
         {
             baseRequest.setHandled(true);
-            HttpOutput out = (HttpOutput) response.getOutputStream();
+            HttpOutput out = (HttpOutput)response.getOutputStream();
 
             // Add interceptor to check aggregation is done
             HttpOutput.Interceptor interceptor = out.getInterceptor();
@@ -749,11 +768,12 @@ public class HttpOutputTest
             out.setWriteListener(new WriteListener()
             {
                 int fill = 0;
+
                 @Override
                 public void onWritePossible() throws IOException
                 {
                     byte[] data = new byte[AggregationChecker.MAX_SIZE];
-                    while(out.isReady())
+                    while (out.isReady())
                     {
                         if (expected.size() >= len)
                         {
@@ -761,7 +781,7 @@ public class HttpOutputTest
                             return;
                         }
 
-                        Arrays.fill(data, (byte)('A' + (fill++%26)));
+                        Arrays.fill(data, (byte)('A' + (fill++ % 26)));
                         expected.write(data);
                         out.write(data);
                     }
@@ -820,11 +840,12 @@ public class HttpOutputTest
     static class AggregateResidueHandler extends AbstractHandler
     {
         ByteArrayOutputStream expected = new ByteArrayOutputStream();
+
         @Override
         public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
         {
             baseRequest.setHandled(true);
-            HttpOutput out = (HttpOutput) response.getOutputStream();
+            HttpOutput out = (HttpOutput)response.getOutputStream();
 
             int bufferSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputBufferSize();
             int commitSize = baseRequest.getHttpChannel().getHttpConfiguration().getOutputAggregationSize();
@@ -855,6 +876,118 @@ public class HttpOutputTest
         }
     }
 
+    @Test
+    public void testPrint() throws Exception
+    {
+        ByteArrayOutputStream bout = new ByteArrayOutputStream();
+        PrintWriter exp = new PrintWriter(bout);
+        _swap.setHandler(new AbstractHandler()
+        {
+            @Override
+            public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+            {
+                baseRequest.setHandled(true);
+                response.setCharacterEncoding("UTF8");
+                HttpOutput out = (HttpOutput)response.getOutputStream();
+
+                exp.print("\u20AC\u0939\uD55C");
+                out.print("\u20AC\u0939\uD55C");
+                exp.print("zero");
+                out.print("zero");
+                exp.print(1);
+                out.print(1);
+                exp.print(2L);
+                out.print(2L);
+                exp.print(3.0F);
+                out.print(3.0F);
+                exp.print('4');
+                out.print('4');
+                exp.print(5.0D);
+                out.print(5.0D);
+                exp.print(true);
+                out.print(true);
+                exp.println("zero");
+                out.println("zero");
+                exp.println(-1);
+                out.println(-1);
+                exp.println(-2L);
+                out.println(-2L);
+                exp.println(-3.0F);
+                out.println(-3.0F);
+                exp.println('4');
+                out.println('4');
+                exp.println(-5.0D);
+                out.println(-5.0D);
+                exp.println(false);
+                out.println(false);
+            }
+        });
+        _swap.getHandler().start();
+        String response = _connector.getResponse("GET / HTTP/1.0\nHost: localhost:80\n\n");
+        assertThat(response, containsString("HTTP/1.1 200 OK"));
+        assertThat(response, containsString(bout.toString()));
+    }
+
+    @Test
+    public void testReset() throws Exception
+    {
+        ByteArrayOutputStream exp = new ByteArrayOutputStream();
+        _swap.setHandler(new AbstractHandler()
+        {
+            @Override
+            public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+            {
+                baseRequest.setHandled(true);
+                HttpOutput out = (HttpOutput)response.getOutputStream();
+                Interceptor interceptor = out.getInterceptor();
+                out.setInterceptor(new Interceptor()
+                {
+                    @Override
+                    public void write(ByteBuffer content, boolean last, Callback callback)
+                    {
+                        interceptor.write(content, last, callback);
+                    }
+
+                    @Override
+                    public Interceptor getNextInterceptor()
+                    {
+                        return interceptor;
+                    }
+
+                    @Override
+                    public boolean isOptimizedForDirectBuffers()
+                    {
+                        return interceptor.isOptimizedForDirectBuffers();
+                    }
+                });
+
+                out.setBufferSize(128);
+                out.println("NOT TO BE SEEN!");
+                out.resetBuffer();
+
+                byte[] data = "TO BE SEEN\n".getBytes(StandardCharsets.ISO_8859_1);
+                exp.write(data);
+                out.write(data);
+
+                out.flush();
+
+                data = "Not reset after flush\n".getBytes(StandardCharsets.ISO_8859_1);
+                exp.write(data);
+                try
+                {
+                    out.resetBuffer();
+                }
+                catch (IllegalStateException e)
+                {
+                    out.write(data);
+                }
+            }
+        });
+        _swap.getHandler().start();
+        String response = _connector.getResponse("GET / HTTP/1.0\nHost: localhost:80\n\n");
+        assertThat(response, containsString("HTTP/1.1 200 OK"));
+        assertThat(response, containsString(exp.toString()));
+    }
 
     private static String toUTF8String(Resource resource)
         throws IOException

--- a/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
+++ b/jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputTest.java
@@ -989,6 +989,38 @@ public class HttpOutputTest
         assertThat(response, containsString(exp.toString()));
     }
 
+    @Test
+    public void testZeroLengthWrite() throws Exception
+    {
+        _swap.setHandler(new AbstractHandler()
+        {
+            @Override
+            public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+            {
+                baseRequest.setHandled(true);
+                response.setContentLength(0);
+                AsyncContext async = request.startAsync();
+                response.getOutputStream().setWriteListener(new WriteListener()
+                {
+                    @Override
+                    public void onWritePossible() throws IOException
+                    {
+                        response.getOutputStream().write(new byte[0]);
+                        async.complete();
+                    }
+
+                    @Override
+                    public void onError(Throwable t)
+                    {
+                    }
+                });
+            }
+        });
+        _swap.getHandler().start();
+        String response = _connector.getResponse("GET / HTTP/1.0\nHost: localhost:80\n\n");
+        assertThat(response, containsString("HTTP/1.1 200 OK"));
+    }
+
     private static String toUTF8String(Resource resource)
         throws IOException
     {

